### PR TITLE
Add num_children C level

### PR DESCRIPTION
**Pull request by GertjanBisschop**
First step towards adding num_children array for Trees. #2274

### CODE REVIEW

File: c/tskit/trees.c [hunk 1]
```diff
@@ -4216,6 +4218,7 @@ tsk_tree_remove_branch(
     parent[c] = TSK_NULL;
     left_sib[c] = TSK_NULL;
     right_sib[c] = TSK_NULL;
+    self->num_children[p]--;
```
Comment (jeromekelleher) [inline c/tskit/trees.c:4221]: We should follow the same pattern above and declare a local ``restrict`` pointer for ``num_children``. This is to help the compiler optimise the code, avoiding some pointer chasing. (Probably doesn't make much difference in practice, but no point in not doing it here.) 

Similarly below.

**Review by jeromekelleher — COMMENTED** [review]
LGTM!

The num_roots stuff should be easy now, it's just tree.num_children[tree.virtual_root]

File: c/tests/test_trees.c [hunk 2]
```diff
@@ -1202,6 +1213,14 @@ test_simplest_unary_records(void)
     CU_ASSERT_EQUAL(tsk_treeseq_get_num_trees(&ts), 1);
     CU_ASSERT_EQUAL(tsk_treeseq_get_num_populations(&ts), 1);
 
+    ret = tsk_tree_init(&t, &ts, 0);
+    CU_ASSERT_EQUAL(ret, 0);
+    ret = tsk_tree_first(&t);
+    CU_ASSERT_EQUAL(t.num_children[2], 1);
+    CU_ASSERT_EQUAL(t.num_children[4], 2);
+    CU_ASSERT_EQUAL(t.num_children[5], tsk_tree_get_num_roots(&t));
```
Comment (jeromekelleher) [inline c/tests/test_trees.c:1221]: This will be a circular test once it's implemented - better to put the actual value in here. Likewise below.

Comment (GertjanBisschop) [reply c/tests/test_trees.c:1221]: Yep. Will change that. Thanks for the comments. Will address them and the Lint error.

**Review by benjeffery — COMMENTED** [review]
LGTM % JK comments

**Comment by jeromekelleher** [comment]
Looks like this much is basically ready to merge. How should we stage the rest of the update @benjeffery? I guess we could hold this off until after 0.5.0 to try and get it shipped and then do the Python updates then? Or should we just push on with the changes now?

**Comment by benjeffery** [comment]
The python side of this should be relatively straightforward right? I'm keen to get 0.5.0 out ASAP, but this is a really nice addition to the release. I think we include this and the edge-diffs then draw a hard line under the release.

**Comment by jeromekelleher** [comment]
Okey dokey - would you mind creating the follow up issues we'd need after this for Python so?

**Comment by GertjanBisschop** [comment]
Cool. I should be able to finish the follow up by the end of the week. 
Shall I squash both commits?

**Comment by jeromekelleher** [comment]
Yep, squash and rebase would be good.

**Review by jeromekelleher — APPROVED** [review]
LGTM!

**Comment by benjeffery** [comment]
I've added a changelog. Will merge.

**Comment by jeromekelleher** [comment]
Looks like you missed updating ``tsk_tree_free`` @GertjanBisschop - this is essential! (Valgrind spotted it though, good ole valgrind)

**Comment by benjeffery** [comment]
@jeromekelleher Did you mean to merge this? Github says you removed the merge label.

**Comment by jeromekelleher** [comment]
I was trying to get mergify to merge - I thought "turning it off and on again" might trigger things.

**Comment by benjeffery** [comment]
Circle CI is missing! Will try re triggering.

**Comment by benjeffery** [comment]
@mergifyio rebase

**Pull request merged**